geo/src/algorithm/triangulate_spade.rs

```diff
@@ -0,0 +1,710 @@
+/// of the vector in direction of its start to close the gap)
+fn remove_lines_by_index<T: SpadeTriangulationFloat>(
+    mut indices: [usize; 2],
+    lines: &mut Vec<Line<T>>,
+) -> [Line<T>; 2] {
+    indices.sort();
+    let [idx0, idx1] = indices;
+    let l1 = lines.remove(idx1);
+    let l0 = lines.remove(idx0);
+    [l0, l1]
+}
+
+/// split lines based on intersection kind:
+///
+/// - intersection point: create 4 new lines from the existing line's endpoints to the intersection
+/// point
+/// - collinear: create 3 new lines (before overlap, overlap, after overlap)
+fn split_lines<T: SpadeTriangulationFloat>(
+    [l0, l1]: [Line<T>; 2],
+    intersection: LineIntersection<T>,
+) -> Vec<Line<T>> {
+    match intersection {
+        LineIntersection::SinglePoint { intersection, .. } => [
+            (l0.start, intersection),
+            (l0.end, intersection),
+            (l1.start, intersection),
+            (l1.end, intersection),
+        ]
+        .map(|(a, b)| Line::new(a, b))
+        .to_vec(),
+        LineIntersection::Collinear { .. } => {
+            let mut points = [l0.start, l0.end, l1.start, l1.end];
+            // sort points by their coordinate values to resolve ambiguities
+            points.sort_by(|a, b| {
+                a.x.partial_cmp(&b.x)
+                    .expect("sorting points by coordinate x failed")
+                    .then_with(|| {
+                        a.y.partial_cmp(&b.y)
+                            .expect("sorting points by coordinate y failed")
+                    })
+            });
+            // since all points are on one line we can just create new lines from consecutive
+            // points after sorting
+            points
+                .windows(2)
+                .map(|win| Line::new(win[0], win[1]))
+                .collect::<Vec<_>>()
+        }
+    }
+}
+
+/// new lines from the `split_lines` function may contain a variety of ill formed lines, this
+/// function cleans all of these cases up
+fn cleanup_filter_lines<T: SpadeTriangulationFloat>(
+    lines_need_check: Vec<Line<T>>,
+    existing_lines: &[Line<T>],
+    known_points: &mut Vec<Coord<T>>,
+) -> Vec<Line<T>> {
+    lines_need_check
+        .into_iter()
+        .map(|mut line| {
+            line.start = snap_or_register_point(line.start, known_points);
+            line.end = snap_or_register_point(line.end, known_points);
+            line
+        })
+        .filter(|l| l.start != l.end)
+        .filter(|l| !existing_lines.contains(l))
+        .filter(|l| !existing_lines.contains(&Line::new(l.end, l.start)))
+        .collect::<Vec<_>>()
+}
+
+/// snap point to the nearest existing point if it's close enough
+fn snap_or_register_point<T: SpadeTriangulationFloat>(
+    point: Coord<T>,
+    known_points: &mut Vec<Coord<T>>,
+) -> Coord<T> {
+    let epsilon_range = <T as std::convert::From<f32>>::from(0.0001);
```

Comment:
Hard coded epsilons like this are tough when we don't know what projection our data is in. 

e.g. in wgs84 this could be ~10 meters. Is that too much? Is there a way we can infer a more appropriate magnitude based on the input?

Reply:
Good point! I really don't like these epsilons anyways, but it was necessary at that time to introduce it. I can take a look at how this influences the triangulation and re-evaluate again. After that I could:

- tune the value again to it's limits
- write more docs to make the constant less magic

I would also be open to make it a top level argument, but then it would probably make the API a bit weird. If tuning the default value fails then I think I'll take this route though.

Reply:
Agree with @michaelkirk .  Is it possible to make this a parameter of the algorithm, so that it's user input?

Reply:
I just made it configurable by adding a `SpadeTriangulationConfig` struct which holds a documented `snap_radius` field and which implements `Default` with the previous epsilon value if you don't care about setting that value yourself.